aptos-move/block-executor/src/modules_and_scripts.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::view::{LatestView, ViewState};
+use aptos_mvhashmap::{types::ShiftedTxnIndex, versioned_module_storage::ModuleStorageReadResult};
+use aptos_types::{
+    executable::{Executable, ModulePath},
+    state_store::{state_key::StateKey, state_value::StateValueMetadata, TStateView},
+    transaction::BlockExecutableTransaction as Transaction,
+    vm::modules::ModuleStorageEntry,
+};
+use aptos_vm_types::module_and_script_storage::{
+    code_storage::AptosCodeStorage, module_storage::AptosModuleStorage,
+};
+use bytes::Bytes;
+use move_binary_format::{
+    errors::{Location, PartialVMResult, VMResult},
+    file_format::CompiledScript,
+    CompiledModule,
+};
+use move_core_types::{account_address::AccountAddress, identifier::IdentStr, metadata::Metadata};
+use move_vm_runtime::{
+    module_cyclic_dependency_error, module_linker_error, CodeStorage, Module, ModuleStorage, Script,
+};
+use std::{collections::HashSet, sync::Arc};
+
+impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> AptosCodeStorage
+    for LatestView<'a, T, S, X>
+{
+}
+
+impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> AptosModuleStorage
+    for LatestView<'a, T, S, X>
+{
+    fn fetch_state_value_metadata(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> PartialVMResult<Option<StateValueMetadata>> {
+        Ok(self
+            .get_module_storage_entry(address, module_name)
+            .map_err(|e| e.to_partial())?
+            .into_module_module_storage_entry_at_idx()
+            .map(|(_, entry)| entry.state_value_metadata().clone()))
+    }
+
+    fn fetch_module_size_by_state_key(
+        &self,
+        state_key: &StateKey,
+    ) -> PartialVMResult<Option<usize>> {
+        // TODO(loader_v2): A very ugly way of converting state keys into generic types.
```

Comment:
🙈

Reply:
I think we need to have AptosModuleStorage generic, and even for module storage: all we need is a generic key which is constructible from address and name. Then we can use ModuleId in MoveVM, StateKey in Aptos, and avpid nested maps. And in future, some index-like compression? Wdyt?

Reply:
Because the trait is not generic, so I guess it should be

Reply:
Fixed in the main PR.